File: backend/src/main/java/com/ultrabms/service/AnnouncementService.java
package com.ultrabms.service;

import com.ultrabms.dto.announcements.*;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.UUID;

/**
 * Service interface for Announcement operations.
 * Handles announcement CRUD, publishing, and lifecycle management.
 *
 * Story 9.2: Internal Announcement Management
 */
public interface AnnouncementService {

    // =================================================================
    // CRUD OPERATIONS
    // =================================================================

    /**
     * Create a new announcement (as DRAFT)
     * AC #50: POST /api/v1/announcements
     *
     * @param dto       AnnouncementCreateDto with announcement data
     * @param createdBy User UUID who is creating the announcement
     * @return Created announcement response DTO
     */
    AnnouncementResponseDto createAnnouncement(AnnouncementCreateDto dto, UUID createdBy);

    /**
     * Get announcement by ID with full details
     * AC #52: GET /api/v1/announcements/{id}
     *
     * @param id Announcement UUID
     * @return Announcement response DTO
     */
    AnnouncementResponseDto getAnnouncementById(UUID id);

    /**
     * Update announcement (DRAFT status only)
     * AC #53: PUT /api/v1/announcements/{id}
     *
     * @param id        Announcement UUID
     * @param dto       AnnouncementUpdateDto with updated data
     * @param updatedBy User UUID who is updating
     * @return Updated announcement response DTO
     */
    AnnouncementResponseDto updateAnnouncement(UUID id, AnnouncementUpdateDto dto, UUID updatedBy);

    /**
     * Delete announcement
     * AC #57: DELETE /api/v1/announcements/{id}
     *
     * @param id        Announcement UUID
     * @param deletedBy User UUID who is deleting
     */
    void deleteAnnouncement(UUID id, UUID deletedBy);

    // =================================================================
    // LIST AND SEARCH
    // =================================================================

    /**
     * Get paginated list of announcements with filters
     * AC #51: GET /api/v1/announcements
     *
     * @param filterDto Filter parameters
     * @param pageable  Pagination parameters
     * @return Page of announcement list DTOs
     */
    Page<AnnouncementListDto> getAnnouncements(AnnouncementFilterDto filterDto, Pageable pageable);

    /**
     * Get active announcements (PUBLISHED and not expired)
     * For Active tab
     *
     * @param pageable Pagination parameters
     * @return Page of active announcements
     */
    Page<AnnouncementListDto> getActiveAnnouncements(Pageable pageable);

    /**
     * Get draft announcements
     * For Drafts tab
     *
     * @param pageable Pagination parameters
     * @return Page of draft announcements
     */
    Page<AnnouncementListDto> getDraftAnnouncements(Pageable pageable);

    /**
     * Get history announcements (EXPIRED and ARCHIVED)
     * For History tab
     *
     * @param pageable Pagination parameters
     * @return Page of history announcements
     */
    Page<AnnouncementListDto> getHistoryAnnouncements(Pageable pageable);

    // =================================================================
    // COPY/DUPLICATE
    // =================================================================

    /**
     * Copy/duplicate an announcement as a new draft
     * AC #54: POST /api/v1/announcements/{id}/copy
     *
     * @param id       Announcement UUID to copy
     * @param copiedBy User UUID who is copying
     * @return New draft announcement with "[Copy] " prefix on title
     */
    AnnouncementResponseDto copyAnnouncement(UUID id, UUID copiedBy);

    // =================================================================
    // PUBLISH/ARCHIVE
    // =================================================================

    /**
     * Publish announcement (sends emails to all active tenants)
     * AC #55: PATCH /api/v1/announcements/{id}/publish
     *
     * @param id          Announcement UUID
     * @param publishedBy User UUID who is publishing
     * @return Published announcement response DTO
     */
    AnnouncementResponseDto publishAnnouncement(UUID id, UUID publishedBy);

    /**
     * Archive announcement
     * AC #56: PATCH /api/v1/announcements/{id}/archive
     *
     * @param id         Announcement UUID
     * @param archivedBy User UUID who is archiving
     * @return Archived announcement response DTO
     */
    AnnouncementResponseDto archiveAnnouncement(UUID id, UUID archivedBy);

    // =================================================================
    // TENANT PORTAL
    // =================================================================

    /**
     * Get active announcements for tenant portal
     * AC #59: GET /api/v1/tenant/announcements
     *
     * @return List of active announcements for tenants
     */
    List<TenantAnnouncementDto> getActiveAnnouncementsForTenants();

    /**
     * Get active announcements for tenant portal with pagination
     *
     * @param pageable Pagination parameters
     * @return Page of tenant announcements
     */
    Page<TenantAnnouncementDto> getActiveAnnouncementsForTenants(Pageable pageable);

    /**
     * Get announcement by ID for tenant view
     *
     * @param id Announcement UUID
     * @return Tenant announcement DTO
     */
    TenantAnnouncementDto getAnnouncementForTenant(UUID id);

    // =================================================================
    // ATTACHMENT
    // =================================================================

    /**
     * Upload attachment for announcement
     *
     * @param id             Announcement UUID
     * @param fileName       Original file name
     * @param contentType    MIME type (must be application/pdf)
     * @param fileContent    File content bytes
     * @param uploadedBy     User UUID who is uploading
     * @return Updated announcement response DTO
     */
    AnnouncementResponseDto uploadAttachment(UUID id, String fileName, String contentType,
                                              byte[] fileContent, UUID uploadedBy);

    /**
     * Delete attachment from announcement
     *
     * @param id        Announcement UUID
     * @param deletedBy User UUID who is deleting
     * @return Updated announcement response DTO
     */
    AnnouncementResponseDto deleteAttachment(UUID id, UUID deletedBy);

    /**
     * Get presigned URL for downloading attachment
     *
     * @param id Announcement UUID
     * @return Presigned download URL
     */
    String getAttachmentDownloadUrl(UUID id);

    // =================================================================
    // SCHEDULED OPERATIONS
    // =================================================================

    /**
     * Expire overdue announcements (called by scheduled job)
     * Updates PUBLISHED announcements past expiresAt to EXPIRED
     *
     * @return Number of announcements expired
     */
    int expireOverdueAnnouncements();

    // =================================================================
    // DASHBOARD
    // =================================================================

    /**
     * Get count of active announcements for dashboard widget
     * AC #60: Dashboard stats endpoint includes announcementCount
     *
     * @return Count of active announcements
     */
    long getActiveAnnouncementCount();

    // =================================================================
    // ANNOUNCEMENT NUMBER GENERATION
    // =================================================================

    /**
     * Generate the next announcement number in format ANN-{YEAR}-{SEQUENCE}
     *
     * @return Generated announcement number
     */
    String generateAnnouncementNumber();
}

File: frontend/src/services/announcement.service.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
/**
 * Announcement API Service
 * Story 9.2: Internal Announcement Management
 *
 * All announcement-related API calls with comprehensive JSDoc documentation
 */

import { apiClient } from '@/lib/api';
import type {
  Announcement,
  AnnouncementFilter,
  CreateAnnouncementRequest,
  UpdateAnnouncementRequest,
  AnnouncementResponse,
  AnnouncementListResponse,
  AnnouncementStats,
  AnnouncementStatsResponse,
  AttachmentDownloadResponse,
  TenantAnnouncement
} from '@/types/announcement';

const ANNOUNCEMENTS_BASE_PATH = '/v1/announcements';
const TENANT_ANNOUNCEMENTS_PATH = '/v1/tenant/announcements';

// ============================================================================
// CREATE ANNOUNCEMENT
// ============================================================================

/**
 * Create a new announcement (as DRAFT)
 *
 * @param data - Announcement creation data (title, message, expiry date)
 *
 * @returns Promise that resolves to the created Announcement with announcementNumber
 *
 * @throws {ValidationException} When validation fails (400)
 * @throws {UnauthorizedException} When JWT token is missing or invalid (401)
 * @throws {ForbiddenException} When user lacks SUPER_ADMIN or ADMIN role (403)
 */
export async function createAnnouncement(data: CreateAnnouncementRequest): Promise<Announcement> {
  const response = await apiClient.post<AnnouncementResponse>(
    ANNOUNCEMENTS_BASE_PATH,
    data
  );
  return response.data.data;
}

// ============================================================================
// LIST ANNOUNCEMENTS
// ============================================================================

/**
 * Get paginated list of announcements with filters
 *
 * @param filters - Optional filters (tab, search, status, date range, etc.)
 *
 * @returns Promise that resolves to paginated list of AnnouncementListItem
 */
export async function getAnnouncements(filters?: AnnouncementFilter): Promise<AnnouncementListResponse> {
  const params: Record<string, any> = {
    page: filters?.page ?? 0,
    size: filters?.size ?? 20,
    sortBy: filters?.sortBy ?? 'createdAt',
    sortDir: filters?.sortDir ?? 'desc'
  };

  // Add filters if provided
  if (filters?.tab) {
    params.tab = filters.tab;
  }
  if (filters?.search) {
    params.search = filters.search;
  }
  if (filters?.status) {
    params.status = filters.status;
  }
  if (filters?.fromDate) {
    params.fromDate = filters.fromDate;
  }
  if (filters?.toDate) {
    params.toDate = filters.toDate;
  }
  if (filters?.createdBy) {
    params.createdBy = filters.createdBy;
  }

  const response = await apiClient.get<AnnouncementListResponse>(
    ANNOUNCEMENTS_BASE_PATH,
    { params }
  );
  return response.data;
}

// ============================================================================
// GET SINGLE ANNOUNCEMENT
// ============================================================================

/**
 * Get announcement by ID with full details
 *
 * @param id - Announcement UUID
 *
 * @returns Promise that resolves to the full Announcement
 */
export async function getAnnouncementById(id: string): Promise<Announcement> {
  const response = await apiClient.get<AnnouncementResponse>(
    `${ANNOUNCEMENTS_BASE_PATH}/${id}`
  );
  return response.data.data;
}

// ============================================================================
// UPDATE ANNOUNCEMENT
// ============================================================================

/**
 * Update announcement (DRAFT status only)
 *
 * @param id - Announcement UUID
 * @param data - Update data (title, message, expiry date)
 *
 * @returns Promise that resolves to the updated Announcement
 *
 * @throws {IllegalStateException} When announcement is not in DRAFT status (400)
 */
export async function updateAnnouncement(
  id: string,
  data: UpdateAnnouncementRequest
): Promise<Announcement> {
  const response = await apiClient.put<AnnouncementResponse>(
    `${ANNOUNCEMENTS_BASE_PATH}/${id}`,
    data
  );
  return response.data.data;
}

// ============================================================================
// DELETE ANNOUNCEMENT
// ============================================================================

/**
 * Delete announcement
 *
 * @param id - Announcement UUID
 */
export async function deleteAnnouncement(id: string): Promise<void> {
  await apiClient.delete(`${ANNOUNCEMENTS_BASE_PATH}/${id}`);
}

// ============================================================================
// COPY ANNOUNCEMENT
// ============================================================================

/**
 * Copy/duplicate an announcement as a new draft
 *
 * @param id - Announcement UUID to copy
 *
 * @returns Promise that resolves to the new draft Announcement with "[Copy] " prefix
 */
export async function copyAnnouncement(id: string): Promise<Announcement> {
  const response = await apiClient.post<AnnouncementResponse>(
    `${ANNOUNCEMENTS_BASE_PATH}/${id}/copy`
  );
  return response.data.data;
}

// ============================================================================
// PUBLISH ANNOUNCEMENT
// ============================================================================

/**
 * Publish announcement (sends emails to all active tenants)
 *
 * @param id - Announcement UUID
 *
 * @returns Promise that resolves to the published Announcement
 *
 * @throws {IllegalStateException} When announcement is not in DRAFT status (400)
 * @throws {IllegalArgumentException} When expiry date is in the past (400)
 */
export async function publishAnnouncement(id: string): Promise<Announcement> {
  const response = await apiClient.patch<AnnouncementResponse>(
    `${ANNOUNCEMENTS_BASE_PATH}/${id}/publish`
  );
  return response.data.data;
}

// ============================================================================
// ARCHIVE ANNOUNCEMENT
// ============================================================================

/**
 * Archive announcement
 *
 * @param id - Announcement UUID
 *
 * @returns Promise that resolves to the archived Announcement
 *
 * @throws {IllegalStateException} When announcement cannot be archived (400)
 */
export async function archiveAnnouncement(id: string): Promise<Announcement> {
  const response = await apiClient.patch<AnnouncementResponse>(
    `${ANNOUNCEMENTS_BASE_PATH}/${id}/archive`
  );
  return response.data.data;
}

// ============================================================================
// ATTACHMENT OPERATIONS
// ============================================================================

/**
 * Upload attachment for announcement (PDF only, max 5MB)
 *
 * @param id - Announcement UUID
 * @param file - PDF file to upload
 *
 * @returns Promise that resolves to the updated Announcement
 *
 * @throws {IllegalStateException} When announcement is not in DRAFT status (400)
 * @throws {IllegalArgumentException} When file is not PDF or exceeds 5MB (400)
 */
export async function uploadAttachment(id: string, file: File): Promise<Announcement> {
  const formData = new FormData();
  formData.append('file', file);

  const response = await apiClient.post<AnnouncementResponse>(
    `${ANNOUNCEMENTS_BASE_PATH}/${id}/attachment`,
    formData,
    {
      headers: {
        'Content-Type': 'multipart/form-data'
      }
    }
  );
  return response.data.data;
}

/**
 * Delete attachment from announcement
 *
 * @param id - Announcement UUID
 *
 * @returns Promise that resolves to the updated Announcement
 */
export async function deleteAttachment(id: string): Promise<Announcement> {
  const response = await apiClient.delete<AnnouncementResponse>(
    `${ANNOUNCEMENTS_BASE_PATH}/${id}/attachment`
  );
  return response.data.data;
}

/**
 * Get presigned URL for downloading attachment
 *
 * @param id - Announcement UUID
 *
 * @returns Promise that resolves to the download URL
 */
export async function getAttachmentDownloadUrl(id: string): Promise<string> {
  const response = await apiClient.get<AttachmentDownloadResponse>(
    `${ANNOUNCEMENTS_BASE_PATH}/${id}/attachment/download`
  );
  return response.data.data.downloadUrl;
}

// ============================================================================
// PDF GENERATION - Story 9.2 AC #35-39
// ============================================================================

/**
 * Download announcement as PDF
 *
 * @param id - Announcement UUID
 *
 * @returns Promise that resolves to Blob containing PDF content
 */
export async function downloadAnnouncementPdf(id: string): Promise<Blob> {
  const response = await apiClient.get(
    `${ANNOUNCEMENTS_BASE_PATH}/${id}/pdf`,
    {
      responseType: 'blob'
    }
  );
  return response.data;
}

/**
 * Open announcement PDF in new tab for print preview
 *
 * @param id - Announcement UUID
 */
export async function openAnnouncementPrintPreview(id: string): Promise<void> {
  const blob = await downloadAnnouncementPdf(id);
  const url = URL.createObjectURL(blob);
  window.open(url, '_blank');
}

/**
 * Download announcement PDF with proper filename
 *
 * @param id - Announcement UUID
 * @param announcementNumber - Announcement number for filename
 */
export async function downloadAnnouncementPdfWithFilename(
  id: string,
  announcementNumber: string
): Promise<void> {
  const blob = await downloadAnnouncementPdf(id);
  const url = URL.createObjectURL(blob);
  const link = document.createElement('a');
  link.href = url;
  link.download = `Announcement-${announcementNumber}.pdf`;
  document.body.appendChild(link);
  link.click();
  document.body.removeChild(link);
  URL.revokeObjectURL(url);
}

// ============================================================================
// STATISTICS
// ============================================================================

/**
 * Get announcement statistics for dashboard
 *
 * @returns Promise that resolves to AnnouncementStats
 */
export async function getAnnouncementStats(): Promise<AnnouncementStats> {
  const response = await apiClient.get<AnnouncementStatsResponse>(
    `${ANNOUNCEMENTS_BASE_PATH}/stats`
  );
  return response.data.data;
}

// ============================================================================
// TENANT PORTAL
// ============================================================================

/**
 * Get active announcements for tenant portal
 *
 * @returns Promise that resolves to list of TenantAnnouncement
 */
export async function getTenantAnnouncements(): Promise<TenantAnnouncement[]> {
  const response = await apiClient.get<{ success: boolean; data: TenantAnnouncement[] }>(
    TENANT_ANNOUNCEMENTS_PATH
  );
  return response.data.data;
}

/**
 * Get single announcement for tenant portal
 *
 * @param id - Announcement UUID
 *
 * @returns Promise that resolves to TenantAnnouncement
 */
export async function getTenantAnnouncementById(id: string): Promise<TenantAnnouncement> {
  const response = await apiClient.get<{ success: boolean; data: TenantAnnouncement }>(
    `${TENANT_ANNOUNCEMENTS_PATH}/${id}`
  );
  return response.data.data;
}

/**
 * Get attachment download URL for tenant portal
 *
 * @param id - Announcement UUID
 *
 * @returns Promise that resolves to the download URL
 */
export async function getTenantAttachmentDownloadUrl(id: string): Promise<string> {
  const response = await apiClient.get<AttachmentDownloadResponse>(
    `${TENANT_ANNOUNCEMENTS_PATH}/${id}/attachment/download`
  );
  return response.data.data.downloadUrl;
}
